### PR TITLE
Model load test TFX component proposal

### PR DESCRIPTION
**Pull request by michalbrys**
Initial commit with the model load test TFX component proposal.

### CODE REVIEW

**Comment by michalbrys** [comment]
@hanneshapke

File: proposals/20210526-model_load_test_component.md [hunk 1]
```diff
@@ -0,0 +1,41 @@
...
+This component could extend a TFX InfraValidator to validate the TensorFlow model serving performance using the load test.
+The load test results will tell us:
+* if the model will meet the prediction time requirements (essential in large scale implementations),
+* if the changes in the new model will make it faster/slower comparing to the baseline model.
+
+## Project Implementation
+The load test component will integrate with TFX using the following approach:
+* The `InfraValidator TFX Pipeline Component` component creates a TensorFlow Serving endpoint.
+* The load test component spins up a pod on the Kubernetes cluster to perform the load test (send a specified volume of the prediction requests to the TensorFlow Serving endpoint).
+* The model serving performance metrics (i.e. average prediction time) and the load test specification (the traffic volume, the network protocol used, etc.) are recorded and stored as an ML Metadata artifact.
```
Comment (hanneshapke) [inline proposals/20210526-model_load_test_component.md:28]: How should the examples for the load test ingested? Via a reference to an example file or via ingesting the output of a 2nd ExampleGen component in the pipeline? The test set should remain constant (ideally).

File: proposals/20210526-model_load_test_component.md [hunk 2]
```diff
@@ -0,0 +1,41 @@
...
+* Produced ML Metadata artifacts could validate prediction speed compared to the baseline model. Based on that, we can decide if we want to proceed or break the TFX pipeline execution.
+
+### Possible limitations:
+* The component will use the model serving endpoint from `InfraValidator TFX Pipeline Component`, it will support only the TensorFlow SavedModel model format, KubeflowDagRunner inside Kubeflow Pipelines and Kubernetes cluster.
+* The prediction time is also related to the infrastructure used for the load test. It means that we should only analyze the model prediction time relatively. The absolute numbers could be hard to reproduce in the production environment.
+
+## Project Dependencies
```
Comment (hanneshapke) [inline proposals/20210526-model_load_test_component.md:35]: Isn't k8s also a dependency here. Especially, the yaml config versions?

File: proposals/20210526-model_load_test_component.md (hunk 1 above)
Comment (michalbrys) [reply proposals/20210526-model_load_test_component.md:28]: Good point, for the results consistency we may need:
- test examples (static)
- ImportExampleGen to read the test examples
- imported test examples as a component input

Comment (chongkong) [reply proposals/20210526-model_load_test_component.md:28]: FYI, ExampleGen's output is `Examples` artifact which typically contains `tf.Example` records. For serving requests we need to convert `tf.Example` to a [`PredictRequest`](https://github.com/tensorflow/serving/blob/master/tensorflow_serving/apis/predict.proto) (or ClassificationRequest or RegressionRequest, but those are rarely used these days) and optionally convert them to json for http request.

Converting `tf.Example` to `PredictRequest` is actually very tricky part because PredictRequest is about how model consumes the example, and it can be literally anything depending on the training code. Currently InfraValidator have a limited support on this auto conversion: it only supports Predict signature that takes serialized tf.Example. In other words, model should include a layer to deserialize it and feed it to the rest of the layers.

You can follow the InfraValidator behavior, or you can directly feed PredictRequest proto or json files using Importer node with custom artifact type.

Comment (michalbrys) [reply proposals/20210526-model_load_test_component.md:28]: @chongkong after doing some tests of `InfraValidator` I think we may follow this behavior.

File: proposals/20210526-model_load_test_component.md [hunk 3]
```diff
@@ -0,0 +1,43 @@
+#### SIG TFX-Addons
+# Project Proposal
+
+**Your name:** Michal Brys
+
+**Your email:** michal.brys@openx.com
+
+**Your company/organization:** OpenX
+
+**Project name:** Model load test TFX Component
```
[the remark this answers is not in the document]
Comment (michalbrys) [reply proposals/20210526-model_load_test_component.md:10]: @chongkong Thanks for the points, I thought about:
(1) Set up workers, send the requests and measure the latency distribution. The distribution metrics (min, max, avg, percentiles) could be published in the output artefact. It may allow us to compare models performance (if we'll keep the workers' number and traffic constant).
(2) From my experience, it's good to have a load test in a separate container to avoid competing for resources. But, it may not be required at this phase if we focus on the model and relative performance metrics.
(3) We may try to use the pubic Docker image for the load test and submit the custom values in params (i.e. https://ghz.sh/docs/usage). Custom Docker image is also an option with less dependency, but more maintaining effort.

**Comment by michalbrys** [comment]
I updated the proposal addressing your comments.

File: proposals/20210526-model_load_test_component.md (hunk 3 above)
[the remark this answers is not in the document]
Comment (chongkong) [reply proposals/20210526-model_load_test_component.md:10]: Sorry for the late follow-up! (Robert reminded me back to this thread)

1. So I assume it is not squeezing the model but test the moderate situation.. Or maybe we can frame it more generally as "any setting is okay; you can configure your model server and testing client setting to your taste, and this component simply provides bridge to that".
2. Agreed that model server should be on the separate container, and actually I was mainly asking about whether the "client" should be on the separate container. If client also runs on the separate container we would have total 3 containers running: 1 for component run, 1 for client, 1 for model server. If client runs on the same container that component runs, we would have total 2 containers: 1 for component + client run, 1 for model server. Unfortunately both options have a challenge which is...
3. (cont.) In either case, it is non-trivial how to bring load test binary (vegeta or ghz) to the container.
    1.  Case 1 (same container for client): Currently we uses `tensorflow/tfx` image for component container. Each component runs with the same image but with the different argument, so that entrypoint decides which component this container would run. If we use this same container as a load test client, such binary should be pre-installed in the container image. But I doubt we will do that for the standard `tensorflow/tfx` image because we cannot manage all dependencies for 3party addons. Instead, we should bake a new image (such as `tfx-addons/loadtest`) that is based on `tensorflow/tfx`, but with vegeta or ghz installed. But TFX team will probably be unable to maintain such container image.
    2.  Case 2 (separate container for client): If there is a separate container image with vegeta/ghz installed, we can use that container image for the load testing client. This is more affordable than case 1 because user can provide external container image as a config, and TFX team doesn't manage any extra container images. It would be even great if vegeta/ghz provides their own container images that TFX can use right away, but I don't know if they have ones. If we go with this option, I would like to ask you about more integration story included in this design.

File: proposals/20210526-model_load_test_component.md [hunk 4]
```diff
@@ -0,0 +1,63 @@
...
+* The model serving performance metrics (i.e. average prediction time) and the load test specification (the traffic volume, the network protocol used, etc.) are recorded and stored as an ML Metadata artifact.
+* Produced ML Metadata artifacts could validate prediction speed compared to the baseline model. Based on that, we can decide if we want to proceed or break the TFX pipeline execution.
+
+### Proposed component spec
+
+INPUTS:
+* `examples: ChannelParameter(type=tfx.types.standard_artifacts.Examples)` - examples used to perform load test. For the reproducibility reason, they should be the same across the multiple load tests, i.e. imported by `tfx.dsl.components.common.importer.Importer` component.
+* `model: ChannelParameter(type=tfx.types.standard_artifacts.Model)` - model used to perform load test. May be output of `Trainer` or `tfx.dsl.components.common.importer.Importer`.
+* `baseline_load_test_results:  ChannelParameter(type=ModelLoadTestResult`) - (optional) the artifact with the load test of the baseline model result. We will use it to compare models speed in the relative way.
```
Comment (chongkong) [inline proposals/20210526-model_load_test_component.md:36]: Let's define what `ModelLoadTestResult` artifact is. Typically artifact have a directory with certain structure (for example fixed filename), and artifact properties (map of str to {number|str|ProtoStruct} values).

My recommendation is to put load test result into a single JSON field artifact property, and leave the directory (artifact.uri) empty unless you have any file-type artifacts:

```python
# I chose the different name here...
class ModelPerformanceMetric(Artifact):
  TYPE_NAME = "ModelPerformanceMetric"
  PROPERTIES = {
    "perf_metric": Property(type=PropertyType.JSON),
    "blessed": Property(type=PropertyType.INT, optional=True)
  }
```

Where perf_metric corresponds to a JSON something like..

```json
{
  "settings": {
    "machine_type": "n1-standard-16",
    "rps": 100
  },
  "result": {
    "inference_time_ms": {
      "average": 14,
      "p50": 16,
      "p75": 22,
      "p90": 29,
      "p99": 44
    }
  }
}
```

File: proposals/20210526-model_load_test_component.md [hunk 5]
```diff
@@ -0,0 +1,63 @@
...
+**Your name:** Michal Brys
+
+**Your email:** michal.brys@openx.com
+
+**Your company/organization:** OpenX
+
+**Project name:** Model load test TFX Component
+
+## Project Description
+A TFX component performs a load test of the exported TensorFlow model because the prediction time may vary on the model type and structure.
```
Comment (chongkong) [inline proposals/20210526-model_load_test_component.md:13]: I just got confused on what's the difference between performance test and load testing, and read https://stackify.com/load-testing-vs-performance-testing-vs-stress-testing/. Can we say that our desired goal is "load testing" not "performance testing"? (because it feels like inference time distribution belong to performance test scope, and thought maybe this component is a PerformanceEvaluator?)
